real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/simulation/SensorPlayer.java

```diff
@@ -57,34 +61,55 @@
     private static final String TERMINATING_VERSION_ID = "STOP";
 
 
-	/**
-	 * Construct a new {@link SensorPlayer} object
-	 * @param filePath {@link String} of the name and location of the log file
-	 */
-	public SensorPlayer(String filePath) {
+    /**
+     * Construct a new {@link SensorPlayer} object
+     * @param filePath {@link String} of the name and location of the log file
+     * @param playbackSpeed the initial playback speed
+     */
+    public SensorPlayer(String filePath, int playbackSpeed) {
 
-		imuPub = new Publisher(NodeChannel.IMU.getMsgPath());
-		gpsPub = new Publisher(NodeChannel.GPS.getMsgPath());
-		encoderPub = new Publisher(NodeChannel.ENCODER.getMsgPath());
-		brakePub = new Publisher(NodeChannel.BRAKE.getMsgPath());
-		steeringPub = new Publisher(NodeChannel.STEERING.getMsgPath());
-		loggingButtonPub = new Publisher(NodeChannel.GUI_LOGGING_BUTTON.getMsgPath());
+        imuPub = new Publisher(NodeChannel.IMU.getMsgPath());
+        gpsPub = new Publisher(NodeChannel.GPS.getMsgPath());
+        encoderPub = new Publisher(NodeChannel.ENCODER.getMsgPath());
+        brakePub = new Publisher(NodeChannel.BRAKE.getMsgPath());
+        steeringPub = new Publisher(NodeChannel.STEERING.getMsgPath());
+        loggingButtonPub = new Publisher(NodeChannel.GUI_LOGGING_BUTTON.getMsgPath());
         logicNotificationPub = new Publisher(NodeChannel.LOGIC_NOTIFICATION.getMsgPath());
 
         translator = new GsonBuilder().create();
 
-		new RobobuggyLogicNotification("initializing the SensorPlayer", RobobuggyMessageLevel.NOTE);
+        new RobobuggyLogicNotification("initializing the SensorPlayer", RobobuggyMessageLevel.NOTE);
+        path = filePath;
+        File f = new File(path);
+        if(!f.exists()) {
+            new RobobuggyLogicNotification("File doesn't exist!", RobobuggyMessageLevel.EXCEPTION);
+        }
+
+        setupPlaybackTrigger();
 
-		path = filePath;
-		File f = new File(path);
-		if(!f.exists()) {
-			new RobobuggyLogicNotification("File doesn't exist!", RobobuggyMessageLevel.EXCEPTION);
-		}
-	}
+        this.playbackSpeed = playbackSpeed;
+    }
 
 
-	@Override
-	public void run() {
+    /**
+     * Sets up the playback trigger - hitting the start button will go through 1 iteration of a log file
+     */
+    public void setupPlaybackTrigger() {
+        new Subscriber(NodeChannel.GUI_LOGGING_BUTTON.getMsgPath(), new MessageListener() {
+            @Override
+            public void actionPerformed(String topicName, Message m) {
+                GuiLoggingButtonMessage message = (GuiLoggingButtonMessage)m;
+                if (message.getLoggingMessage().equals(GuiLoggingButtonMessage.LoggingMessage.START)) {
+                    if (!SensorPlayer.this.isAlive()) {
+                        SensorPlayer.this.start();
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void run() {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [The method run() has an NPath complexity of 202](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2041124211/issues/source?bid=2938665&fileBranchId=3128705#l112)